apis/placement/v1alpha1/eviction_types.go

```diff
@@ -0,0 +1,123 @@
+/*
+Copyright (c) Microsoft Corporation.
+Licensed under the MIT license.
+*/
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +kubebuilder:object:root=true
+// +kubebuilder:resource:scope=Cluster,categories={fleet,fleet-placement}
+// +kubebuilder:subresource:status
+// +kubebuilder:storageversion
+
+// PlacementEviction is an eviction attempt on a specific placement from
+// a ClusterResourcePlacement object; one may use this API to force the removal of specific
+// resources from a cluster.
+//
+// An eviction is a **voluntary** disruption; its execution is subject to the disruption budget
+// linked with the target ClusterResourcePlacement object (if present).
+//
+// Beware that an eviction alone does not guarantee that a placement will not re-appear; i.e.,
+// after an eviction, the Fleet scheduler might still pick the previous target cluster for
+// placement. To prevent this, considering adding proper taints to the target cluster before running
+// an eviction that will exclude it from future placements; this is especially true in scenarios
+// where one would like to perform a cluster replacement.
+//
+// For safety reasons, Fleet will only execute an eviction once; the spec in this object is immutable,
+// and once executed, the object will be ignored afterwards. To trigger another eviction attempt on the
+// same placement from the same ClusterResourcePlacement object, one must re-create (delete and
+// create) the same Eviction object. Note also that an Eviction object will be
+// ignored once it is deemed invalid (e.g., such an object might be targeting a CRP object or
+// a placement that does not exist yet), even if it does become valid later
+// (e.g., the CRP object or the placement appears later). To fix the situation, re-create the
+// Eviction object.
+//
+// Executed evictions might be kept around for a while for auditing purposes; the Fleet controllers might
+// have a TTL set up for such objects and will garbage collect them automatically. For further
+// information, see the Fleet documentation.
+type PlacementEviction struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec is the desired state of the PlacementEviction.
+	//
+	// Note that all fields in the spec are immutable.
+	// +required
+	Spec PlacementEvictionSpec `json:"spec"`
+
+	// Status is the observed state of the PlacementEviction.
+	// +optional
+	Status PlacementEvictionStatus `json:"status,omitempty"`
+}
+
+// PlacementEvictionSpec is the desired state of the PlacementEviction.
+type PlacementEvictionSpec struct {
+	// PlacementReference is the name of the ClusterResourcePlacement object which
```

Comment:
Do we allow to N evictions pointing to the same placement? If not, please add a comment here and state which failures it may cause if there are N evictions for the same placement.

Reply:
Yes, we can have N evictions pointing to the same placement. We evaluate each eviction as we reconcile them.

Reply:
Why do we want to support a behavior different from the k8s experience?

https://kubernetes.io/docs/tasks/run-application/configure-pdb/#arbitrary-controllers-and-selectors

"The eviction API will disallow eviction of any pod covered by multiple PDBs, so most users will want to avoid overlapping selectors. One reasonable use of overlapping PDBs is when pods are being transitioned from one PDB to another."

Reply:
Oh, I think I asked the wrong question. I was trying to ask whether we allow N PDB against one CRP.

Reply:
We don't allow N PDB against one CRP, the current API only applies a PDB if it has the same name as the CRP